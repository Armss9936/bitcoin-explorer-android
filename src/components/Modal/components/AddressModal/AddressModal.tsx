import { View, Modal, StyleSheet, TouchableOpacity } from "react-native";
import { useState } from "react";

import BoxContainerWithText from "src/components/BoxContainerWithText/BoxContainerWithText";
import * as Clipboard from "expo-clipboard";
import ModalHeader from "../ModalHeader/ModalHeader";
import { AddressInfoType, AddressModalProps } from "../../types";
import { initialStateAddress } from "src/mocks/initialStates";
import { getAddressInfo } from "src/api/getData";

export default function AddressModal(props: AddressModalProps) {
  const { keyForSearch, isVisible, handleModalClose } = props;

  const [addressInfo, setAddressInfo] =
    useState<AddressInfoType>(initialStateAddress);

  const copyToClipboard = async (text: string) => {
    await Clipboard.setStringAsync(text);
  };

  return (
    <Modal
      animationType="slide"
      visible={isVisible}
      transparent
      style={styles.modal}
      onShow={async () => {
        const response = await getAddressInfo(keyForSearch);

        const actualBalance =
          response.chain_stats.funded_txo_sum -
          response.chain_stats.spent_txo_sum;

        const formattedAddress: AddressInfoType = {
          address: response.address,
          balance: actualBalance,
          totalAmountReceived: response.chain_stats.funded_txo_sum,
          totalAmountSent: response.chain_stats.spent_txo_sum,
        };

        setAddressInfo(formattedAddress);
      }}
    >
      <View style={{ width: "100%", height: "20%" }}></View>

      <View style={styles.container}>
        <ModalHeader
          title="Endereço"
          handleModalClose={() => {
            handleModalClose();
          }}
        />

        <View style={styles.contentContainer}>
          <TouchableOpacity
            onPress={() => {
              copyToClipboard(keyForSearch);
            }}
          >
            <BoxContainerWithText
              firstText="Endereço"
              secondText={`${addressInfo.address.slice(0, 18)}...`}
            />
          </TouchableOpacity>

          <View>
            <BoxContainerWithText
              firstText="Total Recebido"
              secondText={`${addressInfo.totalAmountReceived / 100000000} BTC`}
              borderStyles={{
                borderBottomEndRadius: 0,
                borderBottomStartRadius: 0,
              }}
            />
            <BoxContainerWithText
              firstText="Total Enviado"
              secondText={`${addressInfo.totalAmountSent / 100000000} BTC`}
              borderStyles={{
                borderTopEndRadius: 0,
                borderTopStartRadius: 0,
                borderBottomEndRadius: 0,
                borderBottomStartRadius: 0,
              }}
            />
            <BoxContainerWithText
              firstText="Saldo"
              secondText={`${addressInfo.balance / 100000000} BTC`}
              borderStyles={{
                borderTopEndRadius: 0,
                borderTopStartRadius: 0,
              }}
            />
          </View>
        </View>
      </View>
    </Modal>
  );
}

const styles = StyleSheet.create({
  modal: {
    flexDirection: "row",
  },
  container: {
    backgroundColor: "#101427",
    height: "80%",
    paddingVertical: 17,
    paddingHorizontal: 10,
  },
  contentContainer: {
    paddingHorizontal: 8,
    marginTop: 35,
    gap: 20,
  },
});
