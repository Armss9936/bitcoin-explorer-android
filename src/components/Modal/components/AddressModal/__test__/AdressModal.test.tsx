/* eslint-disable import/no-extraneous-dependencies */
import renderer from "react-test-renderer";
import {
  initialStateAddress,
  initialStateTransaction,
} from "src/__mocks__/initialStates";
import AddressModal from "../AddressModal";

jest.mock("src/stores/App/useAppStore", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    localization: { t: (key) => key }, // Mock da função t de tradução
    selectedCurrency: "USD", // Moeda mockada
    bitcoinPrice: {
      mempool: { USD: 50000 },
      blockchainInfo: { USD: { buy: 50000 } },
    }, // Preço fictício
  })),
}));

describe("AddressModal", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <AddressModal
          addressInfo={{
            address: "",
            addressData: initialStateAddress,
          }}
          addressTransactions={[
            initialStateTransaction,
            initialStateTransaction,
          ]}
          handleModalClose={() => {}}
          modalType="Address"
          isVisible
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
