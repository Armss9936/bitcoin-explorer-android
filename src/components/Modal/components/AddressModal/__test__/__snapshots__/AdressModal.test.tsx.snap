// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressModal renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  style={
    {
      "flexDirection": "row",
    }
  }
  transparent={true}
  visible={true}
>
  <RCTScrollView
    contentContainerStyle={
      {
        "height": "auto",
        "marginTop": "38%",
        "paddingBottom": "40%",
      }
    }
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          {
            "backgroundColor": "#101427",
            "height": "100%",
            "paddingHorizontal": 10,
            "paddingVertical": 17,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 18,
              }
            }
          >
            address
          </Text>
          <View
            accessibilityLabel="Close Modal"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#D9D9D9",
                "borderRadius": 100,
                "height": 36,
                "justifyContent": "center",
                "opacity": 1,
                "padding": 5,
                "position": "absolute",
                "right": 0,
                "width": 36,
              }
            }
          >
            <svg
              height={20}
              width={20}
            />
          </View>
        </View>
        <View
          style={
            {
              "gap": 20,
              "height": "100%",
              "marginTop": 35,
              "paddingHorizontal": 8,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#1d2133",
                  "borderBottomColor": undefined,
                  "borderBottomEndRadius": undefined,
                  "borderBottomStartRadius": undefined,
                  "borderBottomWidth": 0,
                  "borderRadius": 7,
                  "borderTopEndRadius": undefined,
                  "borderTopStartRadius": undefined,
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "paddingHorizontal": 18,
                  "paddingVertical": 9,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 15,
                  }
                }
              >
                address
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#DF7800",
                      "fontSize": 15,
                    }
                  }
                >
                  ...
                </Text>
              </View>
            </View>
          </View>
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#1d2133",
                  "borderBottomColor": undefined,
                  "borderBottomEndRadius": 0,
                  "borderBottomStartRadius": 0,
                  "borderBottomWidth": 0,
                  "borderRadius": 7,
                  "borderTopEndRadius": undefined,
                  "borderTopStartRadius": undefined,
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "paddingHorizontal": 18,
                  "paddingVertical": 9,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 15,
                  }
                }
              >
                total_received
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#DF7800",
                      "fontSize": 15,
                    }
                  }
                >
                  0 BTC
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#1d2133",
                  "borderBottomColor": undefined,
                  "borderBottomEndRadius": 0,
                  "borderBottomStartRadius": 0,
                  "borderBottomWidth": 0,
                  "borderRadius": 7,
                  "borderTopEndRadius": 0,
                  "borderTopStartRadius": 0,
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "paddingHorizontal": 18,
                  "paddingVertical": 9,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 15,
                  }
                }
              >
                total_sent
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#DF7800",
                      "fontSize": 15,
                    }
                  }
                >
                  0 BTC
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#1d2133",
                  "borderBottomColor": undefined,
                  "borderBottomEndRadius": undefined,
                  "borderBottomStartRadius": undefined,
                  "borderBottomWidth": 0,
                  "borderRadius": 7,
                  "borderTopEndRadius": 0,
                  "borderTopStartRadius": 0,
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "paddingHorizontal": 18,
                  "paddingVertical": 9,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 15,
                  }
                }
              >
                balance
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#DF7800",
                      "fontSize": 15,
                    }
                  }
                >
                  0 BTC
                </Text>
              </View>
            </View>
          </View>
          <View>
            <View
              style={
                {
                  "backgroundColor": "#101427",
                  "gap": 16,
                  "height": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 15,
                    "fontWeight": "600",
                  }
                }
              >
                inputs_and_outputs
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#1d2133",
                    "borderRadius": 8,
                    "flexDirection": "row",
                    "justifyContent": "space-evenly",
                    "maxWidth": "100%",
                    "paddingVertical": 5,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "gap": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 16,
                        }
                      }
                    >
                       ...
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#FFF",
                          "fontSize": 13,
                          "textAlign": "right",
                        }
                      }
                    >
                      0
                       BTC
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 14,
                          "textAlign": "center",
                        }
                      }
                    />
                  </View>
                </View>
                <Text
                  style={
                    {
                      "color": "#D9D9D9",
                      "fontSize": 28,
                      "fontWeight": "800",
                    }
                  }
                >
                  &gt;
                </Text>
                <View
                  style={
                    {
                      "alignItems": "center",
                      "gap": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 16,
                        }
                      }
                    >
                       ...
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#FFF",
                          "fontSize": 13,
                          "textAlign": "right",
                        }
                      }
                    >
                      0
                       BTC
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 14,
                          "textAlign": "center",
                        }
                      }
                    >
                      $ 0,00
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "backgroundColor": "#101427",
                  "gap": 16,
                  "height": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 15,
                    "fontWeight": "600",
                  }
                }
              >
                inputs_and_outputs
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#1d2133",
                    "borderRadius": 8,
                    "flexDirection": "row",
                    "justifyContent": "space-evenly",
                    "maxWidth": "100%",
                    "paddingVertical": 5,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "gap": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 16,
                        }
                      }
                    >
                       ...
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#FFF",
                          "fontSize": 13,
                          "textAlign": "right",
                        }
                      }
                    >
                      0
                       BTC
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 14,
                          "textAlign": "center",
                        }
                      }
                    />
                  </View>
                </View>
                <Text
                  style={
                    {
                      "color": "#D9D9D9",
                      "fontSize": 28,
                      "fontWeight": "800",
                    }
                  }
                >
                  &gt;
                </Text>
                <View
                  style={
                    {
                      "alignItems": "center",
                      "gap": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 16,
                        }
                      }
                    >
                       ...
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#FFF",
                          "fontSize": 13,
                          "textAlign": "right",
                        }
                      }
                    >
                      0
                       BTC
                    </Text>
                    <Text
                      style={
                        {
                          "color": "#DF7800",
                          "fontSize": 14,
                          "textAlign": "center",
                        }
                      }
                    >
                      $ 0,00
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</Modal>
`;
